docroot/modules/custom/mass_entity_usage/mass_entity_usage.module

```diff
@@ -0,0 +1,149 @@
+  // Block tracking of entities with a node source that is unpublished or trash.
+  if (!_mass_entity_usage_tracking_check($source_type, $source_id, $source_vid)) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Find an id/revision_id record in the default field data table.
+ *
+ * @param string $entity_type
+ *   The entity type.
+ * @param int $id
+ *   The entity id.
+ * @param int $vid
+ *   The entity revision id.
+ *
+ * @return bool
+ *   If true, the content should be tracked. If false, do not track.
+ */
+function _mass_entity_usage_tracking_check($entity_type, $id, $vid) {
+  if ($entity_type == 'paragraph') {
+    // Verify the paragraph revision is in the paragraphs_item_field_data table.
+    $query = \Drupal::database()->select('paragraphs_item_field_data', 'pifd');
+    $query->fields('pifd', ['id', 'revision_id']);
+    $query->condition('id', $id);
+    $query->condition('revision_id', $vid);
+    $results = $query->execute()->fetchAll();
+    if (empty($results)) {
+      return FALSE;
+    }
+    // Load the paragraph revision.
+    $paragraph = \Drupal::service('entity_type.manager')->getStorage('paragraph')
+      ->loadByProperties(['id' => $id, 'revision_id' => $vid]);
+    // Get the parent node of the paragraph.
+    $parent_node = Helper::getParentNode(current($paragraph));
+    if (is_null($parent_node)) {
+      return FALSE;
+    }
+    // Check if the parent node should be tracked.
+    _mass_entity_usage_tracking_check('node', $parent_node->id(), $parent_node->getRevisionId());
+  }
+  if ($entity_type == 'node') {
+    // Load the node.
+    $node = \Drupal::entityTypeManager()->getStorage('node')->loadRevision($vid);
+    // Get the moderation state.
+    $content_moderation_state = ContentModerationState::loadFromModeratedEntity($node);
+    if ($content_moderation_state instanceof ContentModerationStateInterface
+      && !$content_moderation_state->get('moderation_state')->isEmpty()) {
+      $state_name = $content_moderation_state->get('moderation_state')->value;
+      // These states should be not be tracked.
+      $skipped_states = [
+        'unpublished',
+        'trash',
+      ];
+      // Don't track sources that are in unpublished or trash state.
+      if (in_array($state_name, $skipped_states)) {
+        return FALSE;
+      }
+    }
+  }
+  return TRUE;
+}
+
+/**
+ * Deletes usage data from revisions that are not the current revision.
+ *
+ * Entity usage also stores data from other revisions than the current one,
+ * but we don't want/need that data, we don't want to see that data in the
+ * entity_usage table either.
+ */
+function mass_entity_usage_keep_only_current_vid($entity) {
+
+  if (!method_exists($entity, 'getRevisionId')) {
+    return;
+  }
+
+  // Only intervene on tracked entities.
+  $to_track = \Drupal::service('config.factory')
+    ->get('entity_usage.settings')
+    ->get('track_enabled_source_entity_types');
```

Comment:
Minor point, but it seems like this should be a method in entity_usage or in custom code. Going to config here is a bit too low level for my taste.

Reply:
I wasn't able to find that method, however, I noticed that the way to get that config on the entity_usage code was doing this: `\Drupal::config('entity_usage.settings')->get('track_enabled_source_entity_types');`, which is the change I'm going to apply. Let me know if that is still correct for you.

Reply:
I guess if they do it that way then its fine for us to do same. No need to make this change unless you want to.